hudi-spark-datasource/hudi-spark/src/test/java/org/apache/hudi/table/upgrade/TestUpgradeDowngradeFixtures.java

```diff
@@ -0,0 +1,633 @@
+      // Timeline layout should change from V1 to V2 during upgrade to V8+
+      Option<TimelineLayoutVersion> layoutVersion = tableConfig.getTimelineLayoutVersion();
+      assertTrue(layoutVersion.isPresent(), "Timeline layout version should be present after upgrade to V8+");
+      assertEquals(TimelineLayoutVersion.LAYOUT_VERSION_2, layoutVersion.get(),
+          "Timeline layout should be upgraded to V2 during upgrade to V8+");
+          
+      // These properties should be added during upgrade to V8+
+      assertTrue(tableConfig.contains(HoodieTableConfig.TIMELINE_PATH),
+          "Timeline path should be added during upgrade to V8+");
+      assertTrue(tableConfig.contains(HoodieTableConfig.RECORD_MERGE_MODE),
+          "Record merge mode should be added during upgrade to V8+");
+      assertTrue(tableConfig.contains(HoodieTableConfig.RECORD_MERGE_STRATEGY_ID),
+          "Record merge strategy ID should be added during upgrade to V8+");
+      assertTrue(tableConfig.contains(HoodieTableConfig.INITIAL_VERSION),
+          "Initial version should be added during upgrade to V8+");
+    }
+  }
+
+  /**
+   * Validate properties removed/changed during downgrade operations.
+   */
+  private void validateDowngradeProperties(HoodieTableVersion fromVersion, HoodieTableVersion toVersion,
+                                         HoodieTableMetaClient metaClient, HoodieTableConfig tableConfig) {
+    LOG.debug("Validating downgrade properties: {} -> {}", fromVersion, toVersion);
+    
+    // Downgrading from V8+ to pre-V8 should remove certain properties
+    if (fromVersion.versionCode() >= HoodieTableVersion.EIGHT.versionCode() 
+        && toVersion.versionCode() < HoodieTableVersion.EIGHT.versionCode()) {
+      // These V8+ properties should be removed or modified during downgrade
+      assertFalse(tableConfig.contains(HoodieTableConfig.TIMELINE_PATH) 
+          && !HoodieTableConfig.TIMELINE_PATH.defaultValue().equals(tableConfig.getString(HoodieTableConfig.TIMELINE_PATH)),
+          "Timeline path should be reset during downgrade from V8+");
+      
+      // Timeline layout should be downgraded to V1
+      Option<TimelineLayoutVersion> layoutVersion = tableConfig.getTimelineLayoutVersion();
+      if (layoutVersion.isPresent()) {
+        assertEquals(TimelineLayoutVersion.LAYOUT_VERSION_1, layoutVersion.get(),
+            "Timeline layout should be downgraded to V1");
+      }
+    }
+  }
+
+  /**
+   * Perform data validation on the table to ensure it remains operational.
+   * This performs a simple read-only validation to verify table accessibility.
+   */
+  private void performDataValidationOnTable(HoodieTableMetaClient metaClient, String stage) {
+    LOG.info("Performing data validation on table {}", stage);
+    
+    try {
+      // Simple read validation to ensure table is accessible
+      long rowCount = performSimpleReadValidation(metaClient);
```

Comment:
Let's do full Dataframe validation using `SparkClientFunctionalTestHarness#areDataframesEqual` by validation column values?

Reply:
Ack, will look into doing this.